decentralized_devstack/provisioning-utils.sh

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env bash
+# Utilities to be included into provisioning scripts.
+
+# Echo all commands being run back to console.
+set -x
+
+# Strictness: fail on errors and undefined variables.
+set -e
+set -o pipefail
+set -u
+
+# Colored text.
+BOLD_GREEN='\033[1;32m'
+BOLD_YELLOW='\033[1;33m'
+BOLD_RED='\033[1;31m'
+NC='\033[0m' # No Color
+
+LAST_MAJOR_LOG_SECONDS="${LAST_MAJOR_LOG_SECONDS:-}"
+
+# TODO document
+log_step(){
+	if [[ -n "$LAST_MAJOR_LOG_SECONDS" ]]; then
+		elapsed=$((SECONDS - LAST_MAJOR_LOG_SECONDS))
+	else
+		elapsed=0
+	fi
+	LAST_MAJOR_LOG_SECONDS="$SECONDS"
+	total="total=$(printf "%03d" "$SECONDS")s"
+	prevstep="prevstep=$(printf "%03d" "$elapsed")s"
+	echo -e "${BOLD_GREEN}[$(date +%T)][${prevstep}][${total}] $*${NC}"
+}
+
+# TODO document
+log_message(){
+	echo -e "${BOLD_GREEN}[$(date +%T)] $*${NC}"
+}
+
+# TODO document
+log_warning(){
+	echo -e "${BOLD_YELLOW}[$(date +%T)] $*${NC}"
+}
+
+# TODO document
+log_error(){
+	echo -e "${BOLD_RED}[$(date +%T)] $*${NC}"
+}
+
+# Execute a shell command in a service's container.
+# Usage:
+#   service_exec SERVICE COMMAND [ARGS...]
+# Examples:
+#   service_exec discovery make migrate
+service_exec(){
+	service="$1"
+	shift
+	command_and_args="$*"
+	docker-compose exec -T "$service" bash -c "$command_and_args"
```

Comment:
```suggestion
	docker-compose exec -T "$service" "$@"
```

I think this would be the correct invocation.